add tidb-upgrade-migration-guide (#20688)
pingcap/docs: tidb-upgrade-migration-guide.md

---
title: Migrate and Upgrade a TiDB Cluster
summary: Learn how to migrate and upgrade a TiDB cluster using BR for full backup and restore, along with TiCDC for incremental data replication.
---

# Migrate and Upgrade a TiDB Cluster

This document describes how to migrate and upgrade a TiDB cluster (also known as a blue-green upgrade) using [BR](/br/backup-and-restore-overview.md) for full backup and restore, along with [TiCDC](/ticdc/ticdc-overview.md) for incremental data replication. This solution uses dual-cluster redundancy and incremental replication to enable smooth traffic switchover and fast rollback, providing a reliable and low-risk upgrade path for critical systems. It is recommended to regularly upgrade the database version to continuously benefit from performance improvements and new features, helping you maintain a secure and efficient database system. The key advantages of this solution include:

> - **Controllable risk**: supports rollback to the old version cluster within minutes, ensuring business continuity.

Reviewer: ```suggestion
- **Controllable risk**: supports rollback to the original cluster within minutes, ensuring business continuity.
```

- **Data integrity**: uses a multi-stage verification mechanism to prevent data loss.
- **Minimal business impact**: requires only a brief maintenance window for the final switchover.

The core workflow for migration and upgrade is as follows:

1. **Pre-upgrade check**: verify cluster status and assess feasibility.
2. **Prepare the new cluster**: create a new cluster from a full backup of the old cluster and upgrade it to the target version.

> 3. **Incremental replication**: establish a forward data replication channel using TiCDC.

Reviewer 2: 把前面这个词，都换成动词开头：Synchronize incremental data

> 4. **Switch and verify**: perform multi-dimensional verification, switch business traffic to the new cluster, and set up a TiCDC rollback channel.
> 5. **Observe status**: maintain the rollback channel. After the observation period, clean up the environment.

Reviewer: ```suggestion
4. **Switch and verify**: perform multi-dimensional verification, switch business traffic to the new cluster, and set up a TiCDC reverse replication channel.
5. **Observe status**: maintain the reverse replication channel. After the observation period, clean up the environment.
```

- **Configuration consistency**: ensure that the [`new_collations_enabled_on_first_bootstrap`](/tidb-configuration-file.md#new_collations_enabled_on_first_bootstrap) configuration is identical between the old and new clusters. Otherwise, BR restore will fail.
- **System table restore**: Use the `--with-sys-table` option during BR restore to recover system table data.

To migrate full data to the new cluster, take the following steps:

1. Perform a full backup on the old cluster:

    ```shell
    tiup br:${cluster_version} backup full --pd ${pd_host}:${pd_port} -s ${backup_location}
    ```

2. Record the TSO of the old cluster for later TiCDC Changefeed creation:

    ```shell
    tiup br:${cluster_version} validate decode --field="end-version" \
    --storage "s3://xxx?access-key=${access-key}&secret-access-key=${secret-access-key}" | tail -n1
    ```

3. Deploy the new cluster:

    ```shell
    tiup cluster deploy ${new_cluster_name} ${cluster_version} tidb-cluster.yaml
    ```

4. Restore the full backup to the new cluster:

    ```shell
    tiup br:${cluster_version} restore full --pd ${pd_host}:${pd_port} -s ${backup_location} --with-sys-table
    ```

> ### 3. Upgrade the new cluster to target version

Reviewer: ```suggestion
### 3. Upgrade the new cluster to the target version
```

        - Because business operations are stopped at this stage, you can use the current TSO.
        - Ensure that `sink-uri` is set to the address of the old cluster to avoid loopback writing risks.

        ```shell
        tiup ctl:${cluster_version} cdc changefeed create --server http://${cdc_host}:${cdc_port} --sink-uri="mysql://${username}:${password}@${tidb_endpoint}:${port}" --config config.toml --start-ts ${tso}

        tiup ctl:${cluster_version} cdc changefeed list --server http://${cdc_host}:${cdc_port}
        ```

8. Redirect business traffic to the new cluster.

9. Monitor the load and operational status of the new cluster using the following Grafana panels:

    - [**TiDB Dashboard > Query Summary**](/grafana-tidb-dashboard.md#query-summary): check the Duration, QPS, and Failed Query OPM metrics.
    - [**TiDB Dashboard > Server**](/grafana-tidb-dashboard.md#server): monitor the **Connection Count** metric to ensure even distribution of connections across nodes.

At this point, business traffic has successfully switched to the new cluster, and the TiCDC reverse replication channel is established.

> ### 3. Emergency rollback

Reviewer: ```suggestion
### 3. Execute emergency rollback
```

The rollback plan is as follows:

- Check data consistency between the new and old clusters regularly to ensure the reverse replication link is operating properly.
- Monitor the system for a specified period, such as one week. If issues occur, switch back to the old cluster.

> - After the monitoring period, remove the reverse replication link and delete the old cluster.

Reviewer: ```suggestion
- After the observation period, remove the reverse replication link and delete the old cluster.
```
